packages/core/src/messagingProtocols/mls/E2eIdentityService/Connection/AcmeServer/AcmeService.ts

```diff
@@ -0,0 +1,199 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+import {AcmeChallenge, AcmeDirectory} from '@wireapp/core-crypto/platforms/web/corecrypto';
+import axios, {AxiosInstance} from 'axios';
+import logdown from 'logdown';
+
+import {
+  GetDirectoryReturnValue,
+  GetInitialNonceReturnValue,
+  PostJoseRequestParams,
+  PostJoseRequestReturnValue,
+} from './AcmeService.types';
+import {
+  AuthorizationResponseData,
+  AuthorizationResponseSchema,
+  DirectoryResponseSchema,
+  NewAccountResponseData,
+  NewAccountResponseSchema,
+  NewOrderResponseData,
+  NewOrderResponseSchema,
+  ResponseHeaderLocation,
+  ResponseHeaderLocationSchema,
+  ResponseHeaderNonce,
+  ResponseHeaderNonceSchema,
+  CheckStatusOfOrderResponseData,
+  CheckStatusOfOrderResponseSchema,
+  DpopChallengeResponseData,
+  DpopChallengeResponseSchema,
+  FinalizeOrderResponseData,
+  FinalizeOrderResponseSchema,
+  OidcChallengeResponseData,
+  OidcChallengeResponseSchema,
+  GetCertificateResponseData,
+  GetCertificateResponseSchema,
+} from './schema';
+
+export class AcmeService {
+  private logger = logdown('@wireapp/core/AcmeService');
+  private readonly axiosInstance: AxiosInstance = axios.create();
+  private readonly url = {
+    DIRECTORY: '/directory',
+  };
+
+  constructor(private discoveryUrl: string) {}
+
+  // ############ Internal Functions ############
+
+  private extractNonce(headers: any): ResponseHeaderNonce['replay-nonce'] {
+    return ResponseHeaderNonceSchema.parse(headers)['replay-nonce'];
+  }
+
+  private extractLocation(headers: any): ResponseHeaderLocation['location'] {
+    return ResponseHeaderLocationSchema.parse(headers).location;
+  }
+
+  private async postJoseRequest<T>({
+    payload,
+    schema,
+    url,
+    errorMessage,
+    shouldGetLocation = false,
+  }: PostJoseRequestParams<T>): PostJoseRequestReturnValue<T> {
+    try {
+      const {data, headers} = await this.axiosInstance.post(url, payload, {
+        headers: {
+          'Content-Type': 'application/jose+json',
+        },
+      });
+      let location = undefined;
+      const nonce = this.extractNonce(headers);
+      if (shouldGetLocation) {
+        location = this.extractLocation(headers);
+      }
+      const accountData = schema.parse(data);
+      return {
+        data: accountData,
+        nonce,
+        location,
+      };
+    } catch (e) {
+      this.logger.error(errorMessage, e);
```

Comment:
Do we already know how we'd handle errors?

Reply:
im wondering if we should return them actually and display them?

Reply:
We could do this, but I dont see the benefit over logging here. The consumer will throw its own errors